RP1 Audio Out

njhollinghurst opened the pull request:
Candidate version of an RP1 Audio Out driver

48kHz stereo only. GPIO 12/13 only.

Now rewritten as an ASOC DAI driver; a generic sound card and "dummy" codec must also be enabled via OF.

njhollinghurst committed:
arm64: dts: Add the Audio Out block to rp1.dtsi

Change the standard rate of PLL_AUDIO_SEC from 192MHz to
153.6MHz to suit audio out.

Declare audio out hardware and give it a named pin control.

Signed-off-by: Nick Hollinghurst <nick.hollinghurst@raspberrypi.com>

njhollinghurst committed:
clk: rp1: Allow audio out to use PLL_AUDIO_SEC; workaround rounding error

Connect PLL_AUDIO_SEC to CLK_AUDIO_OUT, which had been commented out
to avoid interference with I2S: we expect them never to be enabled
at the same time. Work around a rounding error that occurs when the
desired rate is exactly the max but not exactly achievable by the PLL.

Signed-off-by: Nick Hollinghurst <nick.hollinghurst@raspberrypi.com>

njhollinghurst committed:
sound: soc: raspberrypi: RP1 Audio Out driver as an ASOC DAI

Only 48000Hz stereo 16-bit output is currently supported.

It requires some additional OF plumbing to connect it to a
"dummy" codec and generic sound card.

Signed-off-by: Nick Hollinghurst <nick.hollinghurst@raspberrypi.com>

njhollinghurst commented:
@pelwell could you cast an eye over the first two commits please (a54bfe2b4e9091e04b685d19c8565ae4a9f6bc13 572ef48b224199bf8e19afeb8ea54b05ea3a4c6f), to check they don't break I2S or anything else? Thanks.

pelwell commented:
Do you need a card to test with?

njhollinghurst commented:
Successfully tested Justboom DAC Zero and IQaudIO Pi-Codec+ hats at various sample rates.

I could even load and use the rp1aout driver at the same time, but (predictably) I2S stomped on rp1aout's clock when using the (crystal-less) JustBoom hat; they did seem able to coexist with the IQaudIO, though it's not a vital use case.

P33M commented on arch/arm64/boot/dts/broadcom/rp1.dtsi:
```diff
@@ -789,6 +805,16 @@
 				       "gpio24", "gpio25";
 				bias-disable;
 			};
+			rp1_dpi_18bit_cpadhi_gpio1: rp1_dpi_18bit_cpadhi_gpio1 { /* Mode 6 with GPIO1(DEN) but not GPIO0 */
```
dpi? in my audio driver? (it's more likely than you think.)

njhollinghurst replied:
That one was a DO NOT MERGE. That particular DPI mode is incompatible with audio out. It really should have gone on a different branch...

njhollinghurst replied:
Gone now

P33M commented on sound/soc/raspberrypi/rp1_aout.c:
```diff
@@ -0,0 +1,360 @@
+	if (ret)
+		return dev_err_probe(&pdev->dev, ret, "failed to register pcm\n");
+
+	/* TODO: More PM stuff? */
```
I don't think this will have an effect unless you define dev_pm_ops with suspend/resume/idle callbacks.

njhollinghurst replied:
Yes, that is what I was wondering. I suppose it ought to be done.

njhollinghurst replied:
Or maybe there's nothing to be done, as there's very little in this component to consume or save power.

P33M commented on sound/soc/raspberrypi/rp1_aout.c:
```diff
@@ -0,0 +1,360 @@
+	    cmd == SNDRV_PCM_TRIGGER_SUSPEND ||
+	    cmd == SNDRV_PCM_TRIGGER_PAUSE_PUSH) {
+		/* Push a zero sample (assuming DMA has stopped already?) */
+		aout_reg_wr(aout, AUDIO_OUT_SAMPLE_FIFO, 0);
```
Underflow should be handled gracefully by the hardware (the last sample plays forever) - ASoC decouples DMA ops from driver visibility so you may introduce a glitch here if DMA is still running. I don't know if it synchronously drains the DMA buffers or flushes them.

njhollinghurst replied:
In tests it does seem  to use the given value quite reliably -- I'll remove the question mark

It wasn't clear to me whether a step to zero would be more or less objectionable than a nonstandard DC value (perhaps followed by a step later). Maybe 'stop' and 'pause' cases should be treated differently?

njhollinghurst replied:
After a bit more debug: I haven't actually observed a PAUSE_PUSH or SUSPEND -- when I Ctrl-Z `aplay`, it does a STOP. I think I will keep the zero as it seems to work, and may be better than sticking at an arbitrary value.

njhollinghurst commented:
Some unwanted stuff has now been pruned, including the DPI weak.

This PR will likely be abandoned in favour of a more squashed version -- mainly so I can keep the "grungy" version commits, for reference.

We'll need a way to encapsulate enabling Audio Out with cross-Pi compatibility, which may require either more DTS trickery or perhaps firmware support.

P33M commented:
Rebooting with the carrier active doesn't result in a clean shutdown - it goes thump. snd_soc_dai_ops needs a shutdown callback...

njhollinghurst commented:
> Rebooting with the carrier active doesn't result in a clean shutdown - it goes thump. snd_soc_dai_ops needs a shutdown callback...

Ah. Not for me as I have electrolytic capacitors hanging off the output when it goes High-Z. But of course with a buffer it will do that. Will fix in next version.

njhollinghurst commented:
Created a tag https://github.com/njhollinghurst/linux/tree/audio_out_wip with current history.
Merciless squashing will now ensue.

njhollinghurst commented:
Sigh... Something seems to go wrong on about 1 boot in 20.  I get this error:
```
pi@raspberrypi:~ $ aplay -v arpanet_20201206.raw -D hw:CARD=RP1AudioOut -f dat
aplay: main:831: audio open error: Invalid argument
```
The DAI driver sees successful probe, startup and shutdown calls only. In the bad case there is never a call to HW params.

pelwell commented:
For extra credit, create an audremap-pi5 entry in the map that only says bcm2712, so that attempts to load it on other platforms will be blocked.

njhollinghurst committed:
dts: overlays: Enable RP1 Audio Out using audremap-pi5-overlay

Since RP1 Audio Out can only work on GPIOs 12, 13 which would
previously have needed dtoverlay=audremap, overload it both to
enable and pin-map the block (do not enable for other pinouts).

At the same time, generate a default "codec" and "sound card".

Signed-off-by: Nick Hollinghurst <nick.hollinghurst@raspberrypi.com>

njhollinghurst commented:
Sorry for the churn -- I think it's all good now.

- audremap-pi5 is marked as 2712 only
- References to runtime pm removed (as I didn't intend to implement them here/now)
- Taking more care about spiv async register synchronization (at startup and when polling the mute FSM)
- eschewing SET/CLR aliases for superstitious reasons (and because we hardly used and don't need them)
- Reordered PCM and ASOC component registration, to fix a probe-time race condition.
- After moving them around to try to influence probe order, the generic audio nodes are back at top level again.

ak-rex commented:
Will this be backported to 6.6.y?

njhollinghurst commented:
> Will this be backported to 6.6.y?

Wasn't planning to -- we're about to migrate to 6.12 -- but if it applies cleanly I will do so.

popcornmix commented:
@ak-rex the next apt kernel released will be 6.12. There will be no more official 6.6 kernel releases.
Are you wanting this patch in 6.6 for some other reason?

ak-rex commented:
nope, i'll rebase to the newer kernel.
